signer/src/network/libp2p/swarm.rs

```diff
@@ -70,22 +85,66 @@ impl SignerBehavior {
 
         let bootstrap_config = bootstrap::Config::new(local_peer_id)
             .with_initial_delay(config.initial_bootstrap_delay)
-            .add_seed_addresses(config.seed_addresses);
-
+            .add_seed_addresses(&config.seed_addresses);
         let bootstrap = bootstrap::Behavior::new(bootstrap_config);
 
         Ok(Self {
             gossipsub: Self::gossipsub(&keypair)?,
             mdns,
-            kademlia: Self::kademlia(local_peer_id),
+            kademlia,
             ping: Default::default(),
             identify,
             autonat_client,
             autonat_server,
             bootstrap,
+            connection_limits: Self::connection_limits(&config),
         })
     }
 
+    fn connection_limits(config: &SignerSwarmConfig) -> connection_limits::Behaviour {
+        // The number of signers is the number of signers in the set, minus one
+        // for the local signer. This is used to calculate the connection limits.
+        // Note: We use `max(2)` here to ensure that we always have limits
+        // that support at least two signers.
+        let num_signers = config.num_signers.saturating_sub(1).max(2) as u32;
+
+        // Allow `num_signers` * 3 connections to be established at once. This
+        // allows room for one incoming/outgoing connection to each signer, plus
+        // one additional connection per signer as a buffer for protocols such
+        // as autonat.
+        let max_established = num_signers.saturating_mul(3);
+
+        // Allow `num_signers` * 2 _incoming_ connections to be established at
+        // once. This allows room for one incoming connection from each signer,
+        // plus one additional connection per signer as a buffer for protocols
+        // such as autonat.
+        let max_established_incoming = num_signers.saturating_mul(2);
+
+        // Allow for one incoming, one outgoing and one additional connection per
+        // peer.
+        let max_established_per_peer = 3;
+
+        // Allow for `num_signers` _incoming_ connections to be pending at once.
+        // This is conservative, but realistic as we wouldn't expect each signer
+        // to be attempting to open multiple connections to us simultaneously,
+        // even during bootstrapping.
+        let max_pending_incoming = num_signers;
```

Comment:
Since this is before checking peed id a low value could be more detrimental ("external" attempts would count against it). Should we instead use some (high) const?

Reply:
My thinking here is that since these limits are working on accepted sockets, transitioning into the "Upgrade" (negotiation) phase where a malicious peer could purposely hold the connection open for the full timeout duration (now 5s), this helps to prevent possible negotiation-phase resource exhaustion attacks (each negotiation consumes CPU and memory).

I'm not against upping this a bit, but otoh it's just as easy for an attacker to open, say, 65k connections as `num_signers` * 5 or 50.

Note that this value doesn't have any affect on the signers' ability to dial _out_, so as long as their seed peer(s) are reachable during bootstrapping then they will still be able to become well-connected.